From 47529d92fc1d6e6a38b4dc17a234e4653be93b5c Mon Sep 17 00:00:00 2001
From: Florent Pirou <florent.pirou@intel.com>
Date: Wed, 12 Jun 2024 12:39:18 +0200
Subject: [PATCH 18/62] d4xx: add V4L2_CID_LINK_FREQ support MTL DPHY

* enable `v4l2-ctl --set-ctrl v4l2_cid_link_freq=` runtime override
* add D4XX_LINK_FREQ_750MHZ to MTL CSI D-PHY
* add MTL CSI2 D-PHY debug messages

Signed-off-by: Florent Pirou <florent.pirou@intel.com>
---
 drivers/media/i2c/d4xx.c | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/drivers/media/i2c/d4xx.c b/drivers/media/i2c/d4xx.c
index 12dfeb498..d1ef3fbc5 100644
--- a/drivers/media/i2c/d4xx.c
+++ b/drivers/media/i2c/d4xx.c
@@ -211,6 +211,7 @@ enum ds5_mux_pad {
 	if (max9295_write_8(state, addr, buf)) \
 		return -EINVAL; \
 	}
+#define D4XX_LINK_FREQ_750MHZ           750000000ULL
 #define D4XX_LINK_FREQ_360MHZ		360000000ULL
 #define D4XX_LINK_FREQ_300MHZ		300000000ULL
 #define D4XX_LINK_FREQ_288MHZ		288000000ULL
@@ -326,6 +327,7 @@ static const struct hwm_cmd ewb = {
 };
 #ifdef CONFIG_VIDEO_INTEL_IPU6
 static const s64 link_freq_menu_items[] = {
+	D4XX_LINK_FREQ_750MHZ,
 	D4XX_LINK_FREQ_360MHZ,
 	D4XX_LINK_FREQ_300MHZ,
 	D4XX_LINK_FREQ_288MHZ,
@@ -2347,6 +2349,20 @@ static int ds5_s_ctrl(struct v4l2_ctrl *ctrl)
 #endif
 	}
 		break;
+	case V4L2_CID_LINK_FREQ: {
+		if ( sensor && ctrl->p_new.p_u8)
+		{
+		  if (*ctrl->p_new.p_u8 <= (ARRAY_SIZE(link_freq_menu_items) - 1)) {
+			struct v4l2_ctrl *link_freq = state->ctrls.link_freq;
+			dev_info(&state->client->dev,
+				"V4L2_CID_LINK_FREQ modify index to val=%d",
+				 (unsigned int) *ctrl->p_new.p_u8);
+			ret = 0;
+		  }
+		}
+
+	}
+	  break;
 #endif
 	}
 
@@ -2923,7 +2939,7 @@ static const struct v4l2_ctrl_config d4xx_controls_link_freq = {
 	.max = ARRAY_SIZE(link_freq_menu_items) - 1,
 	.min =  0,
 	.step  = 0,
-	.def = 1,
+	.def = 2,    // default D4XX_LINK_FREQ_300MHZ
 	.qmenu_int = link_freq_menu_items,
 };
 
@@ -3594,10 +3610,13 @@ static int ds5_ctrl_init(struct ds5 *state, int sid)
 	}
 #ifdef CONFIG_VIDEO_INTEL_IPU6
 	ctrls->link_freq = v4l2_ctrl_new_custom(hdl, &d4xx_controls_link_freq, sensor);
-
+	/* MTL and RPL/ADL IPU6 CSI-DPHY do NOT share
+	 *  the same default link_freq.
+	 * V4L2_CID_LINK_FREQ must be R/W for udev ot set DPHY platform specific link_freq
+	 * at systemd boottime.
 	if (ctrls->link_freq)
 		ctrls->link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
-
+	*/
 	if (state->aggregated) {
 		d4xx_controls_q_sub_stream.def = NR_OF_DS5_SUB_STREAMS;
 		d4xx_controls_q_sub_stream.min = NR_OF_DS5_SUB_STREAMS;
-- 
2.25.1

